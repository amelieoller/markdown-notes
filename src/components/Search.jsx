import React, { useState, useEffect } from 'react';
import algoliasearch from 'algoliasearch/lite';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';

import Input from '../atoms/Input/Input';

const Search = ({ setSearchResultNotes, placeholderText, clearSearch, border, small }) => {
  const [searchText, setSearchText] = useState('');
  const [hasBeenSearched, setHasBeenSearched] = useState(false);

  const notes = useSelector((state) => state.firestore.ordered.notes);

  useEffect(() => {
    if (!searchText && hasBeenSearched) {
      setHasBeenSearched(false);
      clearSearch();
    }
  }, [searchText]);

  const handleSearch = (text) => {
    setSearchText(text);

    if (text === '') return;
    const lowerCaseText = text.toLowerCase();

    setHasBeenSearched(true);

    // In-App search (only for edittitles)
    const filteredNotes = notes.filter((n) => {
      if (n.textContent) {
        return lowerCaseText.split(' ').every((word) => n.textContent.toLowerCase().includes(word));
      } else {
        return lowerCaseText.split(' ').every((word) => n.title.toLowerCase().includes(word));
      }
    });

    setSearchResultNotes(filteredNotes);
  };

  const clearInput = () => {
    clearSearch();
    setSearchText('');
  };

  return (
    <Input
      label={placeholderText}
      onKeyDown={(e) => e.key === 'Enter' && handleSearch()}
      clearInput={clearInput}
      onChange={(e) => handleSearch(e.target.value)}
      value={searchText}
      border={border}
      small={small}
      showClearButton
    />
  );
};

Search.propTypes = {
  setSearchResultNotes: PropTypes.func.isRequired,
  placeholderText: PropTypes.string,
  clearSearch: PropTypes.func,
  border: PropTypes.bool,
  small: PropTypes.bool,
};

Search.defaultProps = {
  placeholderText: 'Search',
  border: false,
  small: false,
};

export default Search;
