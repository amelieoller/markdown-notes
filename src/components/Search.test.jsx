import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import Search from './Search';

configure({ adapter: new Adapter() });

describe('<Search />', () => {
  it('renders', () => {
    const mockFn = jest.fn();

    shallow(<Search setSearchFilter={mockFn} />);
  });

  it('matches the snapshot', () => {
    const mockFn = jest.fn();
    const wrapper = mount(<Search setSearchFilter={mockFn} />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
