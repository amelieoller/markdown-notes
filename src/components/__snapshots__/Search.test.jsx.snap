// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> matches the snapshot 1`] = `
<Search
  setSearchFilter={[MockFunction]}
>
  <FLoatingLabelInput
    id="search-notes"
    onChange={[Function]}
    onSubmit={[Function]}
    placeholderLabel="Search Notes"
    type="text"
    value=""
  >
    <styled.div
      focused={false}
      id="search-notes"
      onChange={[Function]}
      onSubmit={[Function]}
      placeholderLabel="Search Notes"
      type="text"
      value=""
    >
      <StyledComponent
        focused={false}
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": false,
              "lastClassName": "iwGBpM",
              "rules": Array [
                "
  position: relative;
  display: flex;
  align-items: flex-end;
  padding-right: 1.5rem;
  font-weight: 300;
  margin-top: -10px;

  label {
    position: absolute;
    left: 0;
    top: 17px;
    color: #999;
    z-index: 10;
    transition: transform 0.2s ease-out, font-size 0.2s ease-out;
    line-height: 1;
    max-width: 100%;
    white-space: nowrap;
    overflow: hidden;
    font-size: ",
                [Function],
                ";
    transform: ",
                [Function],
                ";
    cursor: auto;
  }

  input {
    position: relative;
    padding: 16px 0px 3px 0;
    width: 100%;
    outline: 0;
    border: 0;
    box-shadow: 0 1px 0 0 #e5e5e5;
    transition: box-shadow 0.2s ease-out;
    font-size: 16px;
    background: none;
    color: white;
    padding-right: 25px;

    &:focus,
    &.filled {
      box-shadow: 0 1px 0 0 ",
                [Function],
                ";
    }
  }

  .search-icon {
    cursor: pointer;
    z-index: 3;
    margin-left: -17px;
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        id="search-notes"
        onChange={[Function]}
        onSubmit={[Function]}
        placeholderLabel="Search Notes"
        type="text"
        value=""
      >
        <div
          className="sc-bdVaJa iwGBpM"
          id="search-notes"
          onChange={[Function]}
          onSubmit={[Function]}
          type="text"
          value=""
        >
          <input
            className=""
            id="search-notes"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="text"
            value=""
          />
          <label
            htmlFor="search-notes"
            onClick={[Function]}
          >
            Search Notes
          </label>
          <span
            className="search-icon"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <Icon
              className=""
              color="white"
              icon="M992.262 871.396l-242.552-206.294c-25.074-22.566-51.89-32.926-73.552-31.926 57.256-67.068 91.842-154.078 91.842-249.176 0-212.078-171.922-384-384-384-212.076 0-384 171.922-384 384s171.922 384 384 384c95.098 0 182.108-34.586 249.176-91.844-1 21.662 9.36 48.478 31.926 73.552l206.294 242.552c35.322 39.246 93.022 42.554 128.22 7.356s31.892-92.898-7.354-128.22zM384 640c-141.384 0-256-114.616-256-256s114.616-256 256-256 256 114.616 256 256-114.614 256-256 256z"
              size={14}
            >
              <svg
                className=""
                height="14px"
                style={
                  Object {
                    "display": "inline-block",
                    "verticalAlign": "middle",
                  }
                }
                viewBox="0 0 1024 1024"
                width="14px"
              >
                <path
                  d="M992.262 871.396l-242.552-206.294c-25.074-22.566-51.89-32.926-73.552-31.926 57.256-67.068 91.842-154.078 91.842-249.176 0-212.078-171.922-384-384-384-212.076 0-384 171.922-384 384s171.922 384 384 384c95.098 0 182.108-34.586 249.176-91.844-1 21.662 9.36 48.478 31.926 73.552l206.294 242.552c35.322 39.246 93.022 42.554 128.22 7.356s31.892-92.898-7.354-128.22zM384 640c-141.384 0-256-114.616-256-256s114.616-256 256-256 256 114.616 256 256-114.614 256-256 256z"
                  style={
                    Object {
                      "fill": "white",
                    }
                  }
                />
              </svg>
            </Icon>
          </span>
        </div>
      </StyledComponent>
    </styled.div>
  </FLoatingLabelInput>
</Search>
`;
